### PR TITLE
feat: add transform_literal

### PR DESCRIPTION
**Pull request by ZENOTME**
complete #283

### CODE REVIEW

**Comment by ZENOTME** [comment]
cc @liurenjie1024 @Xuanwo @Fokko @marvinlanhenke

**Comment by marvinlanhenke** [comment]
@ZENOTME 
Thank you for this PR. I think I can work with that.

One question though, regarding the input: `&Literal` - I guess we do this, due to the greater flexibility?

However, for #264 I'm working with `BoundPredicates` as input. 

This means I'd still have to expose the underlying `Datum` and its `PrimitiveLiteral` in order to use `fn transform_literal`. Also, when creating a new `UnboundPredicate` I'd have to construct a new `Datum` from the transformed literal.

If we don't need the extra flexibility, changing `fn transform_literal` to 
```rust
fn transform_literal(&self, input: &Datum) -> Result<Option<Datum>>
```
would make the implementation of #264 a lot easier. Perhaps it should be named transform_datum then.

Anyway, thanks again - I'd appreciate your thoughts on this.

**Comment by ZENOTME** [comment]
> @ZENOTME Thank you for this PR. I think I can work with that.
> 
> One question though, regarding the input: `&Literal` - I guess we do this, due to the greater flexibility?
> 
> However, for #264 I'm working with `BoundPredicates` as input.
> 
> This means I'd still have to expose the underlying `Datum` and its `PrimitiveLiteral` in order to use `fn transform_literal`. Also, when creating a new `UnboundPredicate` I'd have to construct a new `Datum` from the transformed literal.
> 
> If we don't need the extra flexibility, changing `fn transform_literal` to
> 
> ```rust
> fn transform_literal(&self, input: &Datum) -> Result<Option<Datum>>
> ```
> 
> would make the implementation of #264 a lot easier. Perhaps it should be named transform_datum then.
> 
> Anyway, thanks again - I'd appreciate your thoughts on this.

For now, `transform` does not rely on the type info, so it can do transform without type info. And `transform_literal` will also be used in manifest metadata, and [manifest store the literal directly](https://github.com/liurenjie1024/iceberg-rust/blob/470e6f08e1291df2bc2ecd22ae9415851beea6ba/crates/iceberg/src/spec/manifest_list.rs#L674). So I think in here what we need is a function extract the reference of literal🤔

**Comment by marvinlanhenke** [comment]
> So I think in here what we need is a function extract the reference of literal🤔

I guess I can do that when implementing #264 since I need this anyway? Other than that thanks again for the PR.

**Comment by ZENOTME** [comment]
> I guess I can do that when implementing https://github.com/apache/iceberg-rust/issues/264 since I need this anyway?

LGTM. Thanks!

File: crates/iceberg/src/transform/temporal.rs [hunk 1]
```diff
@@ -51,12 +61,48 @@ impl TransformFunction for Year {
...
+    }
 }
 
 /// Extract a date or timestamp month, as months from 1970-01-01
 #[derive(Debug)]
 pub struct Month;
 
+impl Month {
+    #[inline]
+    fn timestamp_to_month(timestamp: i64) -> i32 {
```
Comment (liurenjie1024) [inline crates/iceberg/src/transform/temporal.rs:90]: This is incorrect, we should the month is accumulated month, see https://github.com/apache/iceberg/blob/9d42eb44a32ff4a1f86e8427fe023bde1e9652f7/api/src/test/java/org/apache/iceberg/transforms/TestDates.java#L56

**Review by liurenjie1024 — COMMENTED** [review]
Thanks @ZENOTME for this pr, it looks great! I found some small problems and left some comments.

**Review by liurenjie1024 — COMMENTED** [review]
Thanks @ZENOTME This looks great to me! We just need to fix some unnecessary panic.

File: crates/iceberg/src/transform/temporal.rs [hunk 2]
```diff
@@ -51,12 +57,48 @@ impl TransformFunction for Year {
...
+        // date: aaaa-aa-aa
+        // unix epoch date: 1970-01-01
+        // if date > unix epoch date, delta month = (aa - 1) + 12 * (aaaa-1970)
+        // if date < unix epoch date, delta month = (12 - (aa - 1)) + 12 * (1970-aaaa-1)
+        let date = DateTime::from_timestamp_micros(timestamp).unwrap();
+        let unix_epoch_date = DateTime::from_timestamp_micros(0).unwrap();
```
[the remark this answers is not in the document]
Comment (ZENOTME) [reply crates/iceberg/src/transform/temporal.rs:92]: This should be always valid I think.🤔

[the remark this answers is not in the document]
Comment (liurenjie1024) [reply crates/iceberg/src/transform/temporal.rs:92]: Oh, yes, this one is always valid.

**Review by liurenjie1024 — APPROVED** [review]
LGTM, thanks @ZENOTME !

**Pull request merged**